### ISSUE TITLE
session$user is null on rstudio connect after first visit

### ISSUE THREAD

**Issue by davidbrodkeyint** [rstudio/shiny#3218]
### System details

Browser Version: Chrome 86.0.4240.75

Output of `sessionInfo()`:

```
R version 4.0.0 (2020-04-24)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: CentOS Linux 7 (Core)

Matrix products: default
BLAS/LAPACK: /usr/lib64/libopenblasp-r0.3.3.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] shiny_1.5.0   forecast_8.12 DBI_1.1.0     odbc_1.2.2   

loaded via a namespace (and not attached):
 [1] zoo_1.8-8         tidyselect_1.1.0  xfun_0.15         purrr_0.3.4       urca_1.3-0        lattice_0.20-41   colorspace_1.4-1  vctrs_0.3.0       generics_0.0.2    htmltools_0.5.0  
[11] yaml_2.2.1        base64enc_0.1-3   blob_1.2.1        rlang_0.4.6       later_1.0.0       pillar_1.4.4      glue_1.4.1        bit64_0.9-7       TTR_0.24.2        lifecycle_0.2.0  
[21] quantmod_0.4.17   stringr_1.4.0     timeDate_3043.102 munsell_0.5.0     gtable_0.3.0      evaluate_0.14     knitr_1.29        tseries_0.10-47   fastmap_1.0.1     httpuv_1.5.3.1   
[31] lmtest_0.9-37     parallel_4.0.0    curl_4.3          xts_0.12-0        Rcpp_1.0.4.6      xtable_1.8-4      promises_1.1.0    scales_1.1.1      jsonlite_1.6.1    mime_0.9         
[41] bit_1.1-15.2      fracdiff_1.5-1    ggplot2_3.3.1     hms_0.5.3         packrat_0.5.0     digest_0.6.25     stringi_1.4.6     dplyr_1.0.0       grid_4.0.0        quadprog_1.5-8   
[51] tools_4.0.0       magrittr_1.5      tibble_3.0.1      crayon_1.3.4      pkgconfig_2.0.3   ellipsis_0.3.1    rmarkdown_2.3     rstudioapi_0.11   R6_2.4.1          nnet_7.3-13      
[61] nlme_3.1-147      compiler_4.0.0   
```

### Example application *or* steps to reproduce the problem

 

```R
# Example from support.rstudio application example at: https://support.rstudio.com/hc/en-us/articles/360041320233-How-do-I-track-user-activity-within-a-Shiny-application-

library(shiny)

ui <- fluidPage(
    titlePanel("Old Faithful Geyser Data"),
    sidebarLayout(
        sidebarPanel(
            sliderInput("bins",
                        "Number of bins:",
                        min = 1,
                        max = 50,
                        value = 30)
        ),

        mainPanel(
           plotOutput("distPlot"),
           textOutput("sessionUser")
        )
    )
)

server <- function(input, output, session) {
    output$distPlot <- renderPlot({
        # generate bins based on input$bins from ui.R
        x    <- faithful[, 2]
        bins <- seq(min(x), max(x), length.out = input$bins + 1)

        # draw the histogram with the specified number of bins
        hist(x, breaks = bins, col = 'darkgray', border = 'white')
    })
 
    output$sessionUser <- renderText({
      paste("The logged in user is ", session$user, ".")
    })
}

shinyApp(ui = ui, server = server)
```

### Describe the problem in detail
I have a shiny application deployed on R Studio Connect from R Studio Server Pro. The Who can view this application is set to "All users - login required". Users are required to login through company SSO to access application.

I have a very large shiny application and as a part of it I would like to record user activity at certain points. When users press a an action button many actions are triggered one of which is recording who the user is and the values of several inputs. These values are written to an SQL table. My issues comes from the recording the username of the logged in user. I am attempting to use session$user however this value is almost always null. In my attempt to track down the source of the problem I took the example from the article (see above) and deployed that application.

On that deployed application my hope was for the username to appear at the bottom of the application (on connect). Usually the first time I visit the web site after deployment my username correctly appears. However on all subsequent visits to the website including a page reload the area is blank and the username is not present.

**Comment by wch** [comment]
It would help if the you could provide a screenshot with the websocket Network traffic displayed, as shown here. You need to click on `websocket`, then `Messages`.


![image](https://user-images.githubusercontent.com/86978/102558709-ace52280-4093-11eb-8645-6b4be48a8bbc.png)



The important part looks something like:

```
{\"credentials\":\"user\":\"winston\",\"groups\":[]}
```


That is the information that is used to set `session$user` here:

https://github.com/rstudio/shiny/blob/5e1afc61c1b561b4a42ba4feb9428421f1ce60dd/R/shiny.R#L740-L746


I've made a special build of Shiny which logs the information received by the R process running Shiny. You can install it with the following, then restart R and deploy your app:

```R
if (!require(remotes)) install.packages("remotes")
remotes::install_github('rstudio/shiny@wch-log-user')
```

It should result in something like this in the Connect logs:

![image](https://user-images.githubusercontent.com/86978/102558566-57a91100-4093-11eb-92e9-724e8e23fd2c.png)


If that information is being sent by the browser but not being received by this code in the R process running Shiny, then we know that something strange is happening in between.

**Comment by davidbrodkeyint** [comment]
Hello Winston,

Thank you for your response. I opened WebSockets but unfortunately I didn't see a messages tab. It all appears to be all blank I will attach a screenshot my apologies if I was looking at the wrong thing.

I ran your two lines and then restarted R and then redeployed the application. I do see on the Logs a "Credentials string". It says Credentials string: {"user": DBrodkey...   . It is a very very long string which has all my credential information such as groups and permissions. But at the same time on the shiny app the spot for the name is still blank. I will attach a brief screenshot of this but I'm going to cutoff most of it for privacy reasons.

Thank you for your assistance with this.
![shiny_issue_1](https://user-images.githubusercontent.com/76186771/102645288-18240880-4117-11eb-93ed-e43e930bfd07.PNG)
![shiny_issue_2](https://user-images.githubusercontent.com/76186771/102645290-18bc9f00-4117-11eb-8c63-3fa045ea467b.PNG)

**Comment by wch** [comment]
So there is a credentials string but it looks like it is not valid JSON. I wonder if maybe part of it is getting cut off, or if it's getting mangled somehow. Can you copy and paste it into a JSON validator?

You can also check if it is identical to the message sent by the browser, but you'll need to correctly capture it with the Network panel. I think you need to open the Network panel first, and then reload the page. Then you should see the `websocket` in the Name column.

**Comment by wch** [comment]
Oh, I just realized that you don't have the latest version of httpuv. In httpuv 1.5.4, we fixed an issue where HTTP headers could get truncated if they are very long. I suggest you update it with `install.packages('httpuv')` and re-deploy.

https://github.com/rstudio/httpuv/blob/master/NEWS.md#L13-L14

**Comment by davidbrodkeyint** [comment]
I took the credentials string and put it in a JSON validator and it told me it was invalid. I looked at the text more closely and it is getting cutoff (in the logs on connect). Its about 3900 characters (I found this by pasting into word so may not be exact). Here is the structure of it: {"users": "DBrodkey@...","groups":["group_1","group_2",........,"gro

The groups have actual names and the last one is truncated in the middle of its name. If I add "]} to the end of it the JSON then validates.


Just saw your message about the httpuv. I installed the package and redeployed and it looks like its working!

I'll do a bit more testing on my end to make sure it works fully but I think you have found the issue. Thank you very much.

**Comment by wch** [comment]
That's good to hear!

**Comment by davidbrodkeyint** [comment]
I have checked it several times and the simpler application (the one I copied from the r studio example article) continues to work correctly where it displays the logged in name.

Now my actual use case for this functionality involves writing the user login to an SQL table on an action button press and I am still having some issues with that. But, there is a lot going on there so I think it is very likely that the reason for that not working is a different issue.

I'll spend time working on that. But I think we should consider this issue resolved as the functionality works fine on the example app.

Thank you and the whole team for your help with this.
